File: snapshot-middleware/src/parkpow.js
import { fetchWithRetry } from "./utils";

const ORIENTATION_FRONT = "Front";
const ORIENTATION_REAR = "Rear";
const ORIENTATION_UNKNOWN = "Unknown";

export class ParkPowApi {
  constructor(token, sdkUrl = null, retryLimit = 5, retryDelay = 2000) {
    this.retryLimit = retryLimit;
    this.retryDelay = retryDelay;
    if (token === null) {
      throw new Error("ParkPow API token is required for authentication.");
    } else {
      this.token = token;
    }
    if (sdkUrl) {
      this.apiBase = sdkUrl;
    } else {
      this.apiBase = "https://app.parkpow.com";
    }
    // Include API VERSION
    this.apiBase = `${this.apiBase}/api/v1/`;
    // console.debug("Api Base: " + this.apiBase);
  }

  /**
   * Use Log Vehicle API because image is already in base64 from camera
   * https://app.parkpow.com/documentation/#tag/Visits/operation/Create%20Visit%20(Send%20Camera%20Images%20and%20License%20Plate%20Data)
   * @param encodedImage
   * @param result
   * @param timestamp
   * @param cameraId
   * @returns {Promise<*>}
   */
  async logVehicle(encodedImage, result, cameraId, timestamp) {
    console.debug(JSON.stringify(result));
    const endpoint = "log-vehicle/";
    const data = {
      camera: cameraId,
      image: encodedImage,
      results: [result], // TODO Add handling of multiple results
      time: timestamp,
    };
    let init = {
      body: JSON.stringify(data),
      method: "POST",
      headers: {
        "Content-type": "application/json",
        Authorization: "Token " + this.token,
      },
    };
    const url = this.apiBase + endpoint;
    return fetchWithRetry(url, init, this.retryLimit, this.retryDelay).then(
      (res) => res.json(),
    );
  }
}

export const PARKPOW_ORIENTATION_FRONT = ORIENTATION_FRONT;
export const PARKPOW_ORIENTATION_REAR = ORIENTATION_REAR;
export const PARKPOW_ORIENTATION_UNKNOWN = ORIENTATION_UNKNOWN;

File: parkpow/hikvision_lpr/parkpow.py
import logging
import time

import requests

lgr = logging.getLogger(__name__)


class ParkPowApi:
    def __init__(self, token, sdk_url=None):
        if token is None:
            raise Exception("ParkPow TOKEN is required if using Cloud API")
        if sdk_url:
            self.api_base = sdk_url + "/api/v1/"
        else:
            self.api_base = "https://app.parkpow.com/api/v1/"

        lgr.debug(f"Api Base: {self.api_base}")
        self.session = requests.Session()
        self.session.headers = {"Authorization": "Token " + token}

    def log_vehicle(
        self,
        encoded_image,
        license_plate_number,
        confidence,
        camera,
        _datetime,
        plate_coordinates,
    ):
        endpoint = "log-vehicle/"
        p_time = _datetime.strftime("%Y-%m-%d %H:%M:%S.%f%z")
        data = {
            "camera": camera,
            "image": encoded_image,
            "results": [
                {
                    "plate": license_plate_number,
                    "score": confidence,
                    "box": {
                        "xmin": plate_coordinates[0],
                        "ymin": plate_coordinates[1],
                        "xmax": plate_coordinates[0] + plate_coordinates[2],
                        "ymax": plate_coordinates[1] + plate_coordinates[3],
                    },
                }
            ],
            "time": p_time,
        }
        try:
            tries = 0
            while tries < 5:
                response = self.session.post(self.api_base + endpoint, json=data)
                lgr.debug(f"response: {response}")
                if response.status_code < 200 or response.status_code > 300:
                    if response.status_code == 429:
                        tries += 1
                        time.sleep(1)
                    else:
                        lgr.error(response.text)
                        raise Exception("Error logging vehicle")
                else:
                    res_json = response.json()
                    return res_json
        except requests.RequestException as e:
            lgr.error("Error", exc_info=e)
            raise
